File: api/app/ingest/stages/content_gate.py
"""Content gate 스테이지 — PII / 워터마크 / 메신저 감지 (W2 §3.G).

8-stage 파이프라인의 chunk → content_gate → tag_summarize 자리.
저장만 하고 차단·경고 UI 는 W4 (S3 문서 상세) 에서 소비.

검출 항목 (DE-21 정책 + DE-44 결정)
- 주민등록번호: 13자리 + 앞 6자리 YYMMDD 유효성 (DE-21 b 경량)
- 카드번호: 4자리 그룹 4번 패턴 (16~19자리, DE-21 a 단순 매칭)
  - 한국 전화번호 (010-1234-5678 = 11자리) 와 자릿수로 명확히 구분
- 워터마크 키워드: 대외비 · 내부자료 · 보안 · CONFIDENTIAL · INTERNAL (대소문자 무시)
- 메신저 대화: ImageParser 가 ExtractionResult.metadata.vision_type 에 채워둔 값
- 계좌번호: **MVP 에서 제외** (DE-44 — 한국 은행 패턴 너무 다양해 오탐 위험. W3+ 재도입)

저장 스키마 (B-6)
- 문서 수준: `documents.flags.has_pii / has_watermark / third_party` (boolean)
  + `flags.watermark_hits: string[]` (검출 키워드 보존)
- chunk 수준: `chunks.metadata.pii_ranges: [[start, end], ...]`
  + `chunks.metadata.watermark_hits: string[]`
"""

from __future__ import annotations

import dataclasses
import logging
import re
from typing import Any

from app.adapters.parser import ExtractionResult
from app.adapters.vectorstore import ChunkRecord
from app.db import get_supabase_client
from app.ingest.jobs import stage

logger = logging.getLogger(__name__)

_STAGE = "content_gate"

# === PII 정규식 ===
# 주민등록번호: YYMMDD-NXXXXXX (대시 선택)
_RRN_PATTERN = re.compile(r"\b(\d{6})[-\s]?(\d{7})\b")

# 카드번호: 4자리×4그룹 (16~19자리). 마지막 그룹은 4~7자리 (Visa 16, Amex 15·17 등 변형 수용)
# 대시·공백 구분자 허용. 전화번호 (11자리) 와 명확히 분리.
_CARD_PATTERN = re.compile(
    r"\b\d{4}[\s-]?\d{4}[\s-]?\d{4}[\s-]?\d{4,7}\b"
)

# === 워터마크 키워드 ===
_WATERMARK_KEYWORDS = (
    "대외비",
    "내부자료",
    "보안",
    "CONFIDENTIAL",
    "INTERNAL",
)
_WATERMARK_PATTERN = re.compile(
    "|".join(re.escape(k) for k in _WATERMARK_KEYWORDS),
    re.IGNORECASE,
)


def run_content_gate_stage(
    job_id: str,
    *,
    doc_id: str,
    chunks: list[ChunkRecord],
    extraction: ExtractionResult,
) -> tuple[list[ChunkRecord], dict[str, Any]]:
    """chunks 의 metadata 에 pii_ranges/watermark_hits 부착, doc flags 머지.

    반환: (업데이트된 chunks, doc_flags_update dict)
    """
    with stage(job_id, _STAGE):
        has_pii_doc = False
        has_watermark_doc = False
        watermark_hits_doc: set[str] = set()
        updated_chunks: list[ChunkRecord] = []

        for chunk in chunks:
            pii_ranges = _detect_pii(chunk.text)
            watermark_hits = _detect_watermark(chunk.text)

            if pii_ranges:
                has_pii_doc = True
            if watermark_hits:
                has_watermark_doc = True
                watermark_hits_doc.update(watermark_hits)

            new_metadata = dict(chunk.metadata)
            if pii_ranges:
                new_metadata["pii_ranges"] = pii_ranges
            if watermark_hits:
                new_metadata["watermark_hits"] = watermark_hits

            updated_chunks.append(
                dataclasses.replace(chunk, metadata=new_metadata)
            )

        # 메신저 대화 (이미지) — ExtractionResult.metadata.vision_type 참조
        vision_type = (extraction.metadata or {}).get("vision_type")
        third_party = vision_type == "메신저대화"

        flags_update: dict[str, Any] = {
            "has_pii": has_pii_doc,
            "has_watermark": has_watermark_doc,
            "third_party": third_party,
        }
        if watermark_hits_doc:
            flags_update["watermark_hits"] = sorted(watermark_hits_doc)

        _merge_doc_flags(doc_id, flags_update)

        logger.info(
            "content_gate: doc=%s has_pii=%s has_watermark=%s third_party=%s "
            "chunks_with_pii=%d chunks_with_watermark=%d",
            doc_id,
            has_pii_doc,
            has_watermark_doc,
            third_party,
            sum(1 for c in updated_chunks if "pii_ranges" in c.metadata),
            sum(1 for c in updated_chunks if "watermark_hits" in c.metadata),
        )

        return updated_chunks, flags_update


# ---------------------- 내부 헬퍼 ----------------------


def _detect_pii(text: str) -> list[list[int]]:
    """텍스트에서 PII 매칭 위치 → [[start, end], ...] (정렬·중복 제거)."""
    ranges_raw: list[tuple[int, int]] = []

    # 주민번호 — 앞 6자리 YYMMDD 검증 (DE-21 b 경량)
    for m in _RRN_PATTERN.finditer(text):
        if _is_valid_yymmdd(m.group(1)):
            ranges_raw.append((m.start(), m.end()))

    # 카드번호 — 단순 매칭 (DE-21 a)
    for m in _CARD_PATTERN.finditer(text):
        ranges_raw.append((m.start(), m.end()))

    # 정렬 + 중복 제거
    deduped = sorted(set(ranges_raw))
    return [[s, e] for s, e in deduped]


def _is_valid_yymmdd(yymmdd: str) -> bool:
    """주민번호 앞 6자리가 유효 날짜인지 (월 1~12, 일 1~31). 윤년·말일 정밀 체크는 W3+."""
    if len(yymmdd) != 6 or not yymmdd.isdigit():
        return False
    mm = int(yymmdd[2:4])
    dd = int(yymmdd[4:6])
    return 1 <= mm <= 12 and 1 <= dd <= 31


def _detect_watermark(text: str) -> list[str]:
    """워터마크 키워드 매칭 결과 (정렬·중복 제거, 매칭된 표기 보존)."""
    hits: set[str] = set()
    for m in _WATERMARK_PATTERN.finditer(text):
        hits.add(m.group())
    return sorted(hits)


def _merge_doc_flags(doc_id: str, updates: dict[str, Any]) -> None:
    """기존 documents.flags 와 select-then-update 머지 (다른 flags 보존)."""
    client = get_supabase_client()
    resp = (
        client.table("documents")
        .select("flags")
        .eq("id", doc_id)
        .limit(1)
        .execute()
    )
    existing = (resp.data[0].get("flags") if resp.data else None) or {}
    merged = {**existing, **updates}
    client.table("documents").update({"flags": merged}).eq("id", doc_id).execute()

File: supabase/functions/_shared/ingest/content_gate.ts
/**
 * `ingest/stages/content_gate.py` 포팅 — 개인정보·워터마크·메신저 감지.
 *
 * 차단하지 않는다. **표시만 한다** — 문서 상세 화면이 그 값을 읽어 경고를 띄운다.
 *
 * | 저장 위치 | 값 |
 * |---|---|
 * | `documents.flags` | `has_pii` · `has_watermark` · `third_party` · `watermark_hits[]` |
 * | `chunks.metadata` | `pii_ranges: [[start,end], …]` · `watermark_hits[]` |
 *
 * ## 계좌번호는 일부러 안 본다
 * 한국 은행 계좌 형식이 너무 다양해 오탐이 크다는 게 원본의 결정(DE-44)이다.
 *
 * ## 정규식을 그대로 옮기면 안 된다
 * Python `\d` 는 유니코드 `Nd`, `\s` 는 유니코드 공백, `\b` 는 유니코드 단어 경계다.
 * JS 는 셋 다 ASCII 라 아랍-인도 숫자로 쓴 주민번호를 놓친다. `entity_extract.ts` 와
 * 같은 방식으로 풀어 쓴다(`v` 플래그 + 집합 뺄셈).
 *
 * ## 오프셋이 코드포인트다
 * `m.start()` / `m.end()` 는 Python 에서 코드포인트 단위인데 JS 정규식은 UTF-16 이다.
 * 이모지가 앞에 있으면 그만큼 밀린다. `pii_ranges` 는 화면에서 가리기(마스킹)에 쓰이는
 * 값이라 밀리면 엉뚱한 자리를 가린다 — 변환해서 넣는다.
 */

import { PY_SP } from "../search/pystr.ts";
import { PY_WORD_CLASS, pyIsDigit } from "../pychar.ts";
import { pyInt } from "../pynum.ts";
import type { ChunkRecord } from "./chunk_records.ts";

/** Python `\b` — `PY_WORD_CLASS` 의 경계. `v` 플래그가 있어야 집합 뺄셈이 산다. */
const B = `(?:(?<=${PY_WORD_CLASS})(?!${PY_WORD_CLASS})|(?<!${PY_WORD_CLASS})(?=${PY_WORD_CLASS}))`;
const D = "\\p{Nd}";

/** 주민등록번호 — `YYMMDD[-공백]?NXXXXXX`. 앞 6 자리는 날짜 검증까지 한다. */
// `v` 플래그에서 `-` 는 문자 클래스 안의 예약 문자다 — 반드시 이스케이프한다.
// 안 하면 정규식 생성 자체가 SyntaxError 로 죽는다(실제로 여기서 걸렸다).
const RRN_RE = new RegExp(`${B}(${D}{6})[\\-${PY_SP}]?(${D}{7})${B}`, "gv");

/**
 * 카드번호 — 4 자리 × 3 그룹 + 마지막 4~7 자리(16~19 자리).
 *
 * 한국 휴대폰 번호(11 자리)와 자릿수로 갈린다 — 그래서 마지막 그룹 하한이 4 다.
 */
const CARD_RE = new RegExp(
  `${B}${D}{4}[${PY_SP}\\-]?${D}{4}[${PY_SP}\\-]?${D}{4}[${PY_SP}\\-]?${D}{4,7}${B}`,
  "gv",
);

export const WATERMARK_KEYWORDS = [
  "대외비",
  "내부자료",
  "보안",
  "CONFIDENTIAL",
  "INTERNAL",
] as const;

/** `re.IGNORECASE` 대응. 한글은 대소문자가 없어 영문 두 개만 실제로 영향받는다. */
const WATERMARK_RE = new RegExp(WATERMARK_KEYWORDS.join("|"), "giu");

/** UTF-16 인덱스 → 코드포인트 인덱스. Python 오프셋과 맞추려면 필요하다. */
function cpIndexMap(text: string): number[] {
  // `map[utf16Index] = codePointIndex`. 서로게이트 뒤쪽 자리도 채워 둔다.
  const map = new Array<number>(text.length + 1);
  let cp = 0;
  let i = 0;
  for (const ch of text) {
    for (let k = 0; k < ch.length; k++) map[i + k] = cp;
    i += ch.length;
    cp++;
  }
  map[text.length] = cp;
  return map;
}

/** 원본 `_is_valid_yymmdd` — 월 1~12, 일 1~31. 윤년·말일은 안 본다. */
export function isValidYymmdd(yymmdd: string): boolean {
  // 원본은 `len()` 과 `isdigit()` 을 본다. 정규식이 이미 `Nd` 6 자리를 보장해서
  // `isdigit()` 은 그 경로에서 항상 참이지만, **함수 자체를 원본과 같게 둔다** —
  // 안 그러면 `"90 101"` 같은 입력에서 갈린다(대조가 실제로 잡았다).
  // `str.isdigit()` 은 `Nd` 보다 넓다(위첨자 ² 등) — `pyIsDigit` 이 그 집합이다.
  let n = 0;
  for (const ch of yymmdd) {
    if (!pyIsDigit(ch)) return false;
    n++;
  }
  if (n !== 6) return false;
  const mm = pyInt(yymmdd.slice(2, 4));
  const dd = pyInt(yymmdd.slice(4, 6));
  if (mm === null || dd === null) return false;
  return mm >= 1 && mm <= 12 && dd >= 1 && dd <= 31;
}

/** 원본 `_detect_pii` — 정렬·중복 제거된 `[start, end]` 목록(코드포인트 기준). */
export function detectPii(text: string): [number, number][] {
  const map = cpIndexMap(text);
  const raw: [number, number][] = [];

  RRN_RE.lastIndex = 0;
  for (const m of text.matchAll(RRN_RE)) {
    if (isValidYymmdd(m[1])) {
      raw.push([map[m.index!], map[m.index! + m[0].length]]);
    }
  }
  CARD_RE.lastIndex = 0;
  for (const m of text.matchAll(CARD_RE)) {
    raw.push([map[m.index!], map[m.index! + m[0].length]]);
  }

  // `sorted(set(...))` — 튜플 정렬은 start, 그다음 end 다.
  const seen = new Set<string>();
  const out: [number, number][] = [];
  for (const r of raw) {
    const k = `${r[0]},${r[1]}`;
    if (seen.has(k)) continue;
    seen.add(k);
    out.push(r);
  }
  out.sort((a, b) => (a[0] - b[0]) || (a[1] - b[1]));
  return out;
}

/** 코드포인트 순 비교 — JS 기본 정렬은 UTF-16 단위라 astral 에서 갈린다. */
function cpCompare(a: string, b: string): number {
  const ai = [...a];
  const bi = [...b];
  for (let i = 0; i < Math.min(ai.length, bi.length); i++) {
    const d = ai[i].codePointAt(0)! - bi[i].codePointAt(0)!;
    if (d !== 0) return d;
  }
  return ai.length - bi.length;
}

/** 원본 `_detect_watermark` — 매칭된 **표기 그대로** 보존한다(대소문자 포함). */
export function detectWatermark(text: string): string[] {
  const hits = new Set<string>();
  WATERMARK_RE.lastIndex = 0;
  for (const m of text.matchAll(WATERMARK_RE)) hits.add(m[0]);
  return [...hits].sort(cpCompare);
}

export interface ContentGateResult {
  chunks: ChunkRecord[];
  flagsUpdate: Record<string, unknown>;
  chunksWithPii: number;
  chunksWithWatermark: number;
}

/**
 * 원본 `run_content_gate_stage` — 청크 metadata 를 채우고 문서 flags 를 만든다.
 *
 * DB 쓰기는 하지 않는다. 호출자가 `flagsUpdate` 를 기존 flags 에 머지한다.
 *
 * `visionType` 은 `ExtractionResult.metadata.vision_type` 이다. 단독 이미지 업로드에서만
 * 채워진다 — **스캔 PDF 경로는 원본이 그 값을 안 넘긴다**(`_reroute_pdf_to_image` 가
 * `ExtractionResult` 에 metadata 를 안 준다). 그래서 PDF 는 항상 `third_party=false` 다.
 */
export function runContentGateStage(opts: {
  chunks: ChunkRecord[];
  visionType?: unknown;
}): ContentGateResult {
  let hasPii = false;
  let hasWatermark = false;
  const watermarkHitsDoc = new Set<string>();
  const updated: ChunkRecord[] = [];

  for (const chunk of opts.chunks) {
    const piiRanges = detectPii(chunk.text);
    const watermarkHits = detectWatermark(chunk.text);
    if (piiRanges.length > 0) hasPii = true;
    if (watermarkHits.length > 0) {
      hasWatermark = true;
      for (const h of watermarkHits) watermarkHitsDoc.add(h);
    }
    const metadata = { ...chunk.metadata };
    if (piiRanges.length > 0) metadata["pii_ranges"] = piiRanges;
    if (watermarkHits.length > 0) metadata["watermark_hits"] = watermarkHits;
    updated.push({ ...chunk, metadata });
  }

  const flagsUpdate: Record<string, unknown> = {
    has_pii: hasPii,
    has_watermark: hasWatermark,
    third_party: opts.visionType === "메신저대화",
  };
  if (watermarkHitsDoc.size > 0) {
    flagsUpdate["watermark_hits"] = [...watermarkHitsDoc].sort(cpCompare);
  }

  return {
    chunks: updated,
    flagsUpdate,
    chunksWithPii: updated.filter((c) => "pii_ranges" in c.metadata).length,
    chunksWithWatermark: updated.filter((c) => "watermark_hits" in c.metadata).length,
  };
}
